Add use case for anomaly detection with Data Prepper (#6140)
opensearch-project/documentation-website: _data-prepper/common-use-cases/anomaly-detection.md

---
layout: default

> title: Anomaly detection with Data Prepper

Reviewer: Can we drop the " with Data Prepper" part to shorten this title?

parent: Common use cases
nav_order: 30
---

# Anomaly detection

> You can use Data Prepper to train models and generate anomalies in near real-time on time-series aggregated events. You can generate anomalies either on events generated within the pipeline, or on events coming directly into the pipeline, like OpenTelemetry metrics. You can feed these tumbling window aggregated time-series events to the [`anomaly_detector` processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/anomaly-detector/), which trains a model and generates anomalies with a grade score. Then write the anomalies to a separate index to create document monitors and trigger fast alerting.

Reviewer 2: ```suggestion
You can use Data Prepper to train models and generate anomalies in near real time on time-series aggregated events. You can generate anomalies either on events generated within the pipeline or on events coming directly into the pipeline, like OpenTelemetry metrics. You can feed these tumbling window aggregated time-series events to the [`anomaly_detector` processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/anomaly-detector/), which trains a model and generates anomalies with a grade score. Then write the anomalies to a separate index to create document monitors and trigger fast alerting.
```

Reviewer: > Then write the anomalies to a separate index to create document monitors and trigger fast alerting.

The subject is unclear in my view. 

Maybe:

> You can configure your pipeline to write the anomalies to a separate index to create document monitors and trigger fast alerting.

## Metrics from logs 

> The following pipeline receives logs using an HTTP source like FluentBit, extracts important values from the logs by matching the value in the `log` key against the grok common Apache log pattern, and then forwards the grokked logs to both the `log-to-metrics-pipeline sub-pipeline` and to an OpenSearch index named `logs`.

Reviewer: What you have as "common Apache log pattern" is called "Apache Common Log Format".

We may want to link it to this page: https://httpd.apache.org/docs/2.4/logs.html#accesslog

> The following pipeline receives logs using an HTTP source like FluentBit, extracts important values from the logs by matching the value in the `log` key against the grok [Apache Common Log Format](https://httpd.apache.org/docs/2.4/logs.html#accesslog), and then forwards the grokked logs to both the `log-to-metrics-pipeline sub-pipeline` and to an OpenSearch index named `logs`.

Reviewer 2: ```suggestion
The following pipeline receives logs using an HTTP source like FluentBit, extracts important values from the logs by matching the value in the `log` key against the grok [Apache Common Log Format](https://httpd.apache.org/docs/2.4/logs.html#accesslog), and then forwards the grokked logs to both the `log-to-metrics-pipeline sub-pipeline` and an OpenSearch index named `logs`.
```

Reviewer: > `log-to-metrics-pipeline sub-pipeline`

Change this to:

> `log-to-metrics-pipeline` pipeline

> The following pipeline receives logs using an HTTP source like FluentBit, extracts important values from the logs by matching the value in the `log` key against the grok [Apache Common Log Format](https://httpd.apache.org/docs/2.4/logs.html#accesslog), and then forwards the grokked logs to both the `log-to-metrics-pipeline sub-pipeline` and to an OpenSearch index named `logs`.

Reviewer 2: First sentence: "from" instead of "using"? Should "Grok" be capitalized?

Author: USe "grok" lowercase when it's not a proper name. The proper name "Grok" is X's AI model. The OpenSearch processor is named `grok`. When we use the processor name, I think we should put it in code format so that it's clear to the reader when we're referring to the OpenSearch processor. Otherwise, lowercase "grok" when it's commonly used.

> The `log-to-metrics-pipeline` sub-pipeline receives the grokked logs from the `apache-log-pipeline-with-metrics sub-pipeline`, aggregates them, and derives histogram metrics based on the values in the `clientip` and `request` keys. It then sends the histogram metrics to an OpenSearch index named `histogram_metrics`, as well as to the `log-to-metrics-anomaly-detector` sub-pipeline.

Reviewer 2: ```suggestion
The `log-to-metrics-pipeline` sub-pipeline receives the grokked logs from the `apache-log-pipeline-with-metrics sub-pipeline`, aggregates them, and derives histogram metrics based on the values in the `clientip` and `request` keys. It then sends the histogram metrics to an OpenSearch index named `histogram_metrics` as well as to the `log-to-metrics-anomaly-detector` sub-pipeline.
```

Reviewer: Data Prepper doesn't have a concept of "sub-pipelines." Let's use "pipeline" instead.

Also,

> `apache-log-pipeline-with-metrics sub-pipeline`

to

> `apache-log-pipeline-with-metrics` pipeline

> The `log-to-metrics-anomaly-detector-pipeline` sub-pipeline receives the aggregated histogram metrics from the `log-to-metrics-pipeline` sub-pipeline and sends them to the `anomaly_detector` processor to detect anomalies using the Random Cut Forest algorithm. If it detects anomalies, it sends them to an OpenSearch index named `log-metric-anomalies`.

Reviewer 2: ```suggestion
The `log-to-metrics-anomaly-detector-pipeline` sub-pipeline receives the aggregated histogram metrics from the `log-to-metrics-pipeline` sub-pipeline and sends them to the `anomaly_detector` processor to detect anomalies by using the Random Cut Forest algorithm. If it detects anomalies, it sends them to an OpenSearch index named `log-metric-anomalies`.
```

> The `log-to-metrics-anomaly-detector-pipeline` sub-pipeline receives the aggregated histogram metrics from the `log-to-metrics-pipeline` sub-pipeline and sends them to the `anomaly_detector` processor to detect anomalies using the Random Cut Forest algorithm. If it detects anomalies, it sends them to an OpenSearch index named `log-metric-anomalies`.

Reviewer 2: In the last sentence, can we replace the first instance of "it" with the noun? "If the algorithm detects..."?

        index: "logs"
    - pipeline:
        name: "log-to-metrics-pipeline"

log-to-metrics-pipeline:
  source:
    pipeline:
      name: "apache-log-pipeline-with-metrics"
  processor:
    - aggregate:
        # Specify the required identification keys
        identification_keys: ["clientip", "request"]
        action:
          histogram:
            # Specify the appropriate values for each the following fields
            key: "bytes"
            record_minmax: true
            units: "bytes"
            buckets: [0, 25000000, 50000000, 75000000, 100000000]
        # Pick the required aggregation period
        group_duration: "30s"
  sink:
    - opensearch:
        ...
        index: "histogram_metrics"
    - pipeline:
        name: "log-to-metrics-anomaly-detector-pipeline"

log-to-metrics-anomaly-detector-pipeline:
  source:
    pipeline:
      name: "log-to-metrics-pipeline"
  processor:
    - anomaly_detector:
        # Specify the key on which to run anomaly detection
        keys: [ "bytes" ]
        mode:
          random_cut_forest:
  sink:
    - opensearch:
        ...
        index: "log-metric-anomalies"
```
{% include copy-curl.html %}

## Metrics from traces

> You can derive metrics from traces and find anomalies in these generated metrics. In this example, the `entry-pipeline` sub-pipeline receives trace data from the OpenTelemetry Collector and forwards it to the following sub-pipelines:

Reviewer 2: Can "generated" be safely removed here?

> You can derive metrics from traces and find anomalies in these generated metrics. In this example, the `entry-pipeline` sub-pipeline receives trace data from the OpenTelemetry Collector and forwards it to the following sub-pipelines:

Reviewer 2: ```suggestion
You can derive metrics from traces and find anomalies in those generated metrics. In this example, the `entry-pipeline` sub-pipeline receives trace data from the OpenTelemetry Collector and forwards it to the following sub-pipelines:
```

- `span-pipeline` – Extracts the raw spans from the traces. It sends the raw spans to any indexes OpenSearch prefixed with `otel-v1-apm-span`.

> - `service-map-pipeline` – Aggregates and analyzes it to create documents that represent connections between services. It sends these documents to an OpenSearch index named `otel-v1-apm-service-map`. You can then see a visualization of the service map through the [Trace Analytics]({{site.url}}{{site.baseurl}}/observing-your-data/trace/index/) plugin for OpenSearch Dashboards.

Reviewer 2: Can we replace the first instance of "it" with the noun? Aggregates and analyzes what?

> - `trace-to-metrics-pipeline` - Aggregates and derives histogram metrics from the traces based on the value of the `serviceName`. It then sends the derived metrics to an OpenSearch index named `metrics_for_traces` and to the `trace-to-metrics-anomaly-detector-pipeline` sub-pipeline.

Reviewer 2: Confirm en dash instead of hyphen.

> The `trace-to-metrics-anomaly-detector-pipeline` sub-pipeline receives the aggregated histogram metrics from the `trace-to-metrics-pipeline` and sends them to the `anomaly_detector` processor to detect anomalies using the Random Cut Forest algorithm. If it detects any anomalies, it sends them to an OpenSearch index named `trace-metric-anomalies`.

Reviewer 2: ```suggestion
The `trace-to-metrics-anomaly-detector-pipeline` sub-pipeline receives the aggregated histogram metrics from the `trace-to-metrics-pipeline` and sends them to the `anomaly_detector` processor to detect anomalies by using the Random Cut Forest algorithm. If it detects any anomalies, it sends them to an OpenSearch index named `trace-metric-anomalies`.
```

> The `trace-to-metrics-anomaly-detector-pipeline` sub-pipeline receives the aggregated histogram metrics from the `trace-to-metrics-pipeline` and sends them to the `anomaly_detector` processor to detect anomalies using the Random Cut Forest algorithm. If it detects any anomalies, it sends them to an OpenSearch index named `trace-metric-anomalies`.

Reviewer 2: In the last sentence, can we replace the first instance of "it" with the noun? "If the algorithm detects..."?

        index_type: "trace-analytics-service-map"

trace-to-metrics-pipeline:
  source:
    pipeline:
      name: "entry-pipeline"
  processor:
    - aggregate:
        # Pick the required identification keys
        identification_keys: ["serviceName"]
        action:
          histogram:
            # Pick the appropriate values for each the following fields
            key: "durationInNanos"
            record_minmax: true
            units: "seconds"
            buckets: [0, 10000000, 50000000, 100000000]
        # Pick the required aggregation period
        group_duration: "30s"
  sink:
    - opensearch:
        ...
        index: "metrics_for_traces"
    - pipeline:
        name: "trace-to-metrics-anomaly-detector-pipeline"

trace-to-metrics-anomaly-detector-pipeline:
  source:
    pipeline:
      name: "trace-to-metrics-pipeline"
  processor:
    - anomaly_detector:
        # Below Key will find anomalies in the max value of histogram generated for durationInNanos.
        keys: [ "max" ]
        mode:
          random_cut_forest:
  sink:
    - opensearch:
        ...
        index: "trace-metric-anomalies"
```
{% include copy-curl.html %}

## OpenTelemetry metrics

> You can create a pipeline that receives OpenTelemetry metrics and detects anomalies in these metrics. In this example, `entry-pipeline` receives metrics data from the OpenTelemetry Collector. If a metric is of type `GAUGE` and the name of the metric is `totalApiBytesSent`, the processor sends it to the `ad-pipeline` sub-pipeline.

Reviewer 2: ```suggestion
You can create a pipeline that receives OpenTelemetry metrics and detects anomalies in those metrics. In this example, `entry-pipeline` receives metrics from the OpenTelemetry Collector. If a metric is of type `GAUGE` and the name of the metric is `totalApiBytesSent`, the processor sends it to the `ad-pipeline` sub-pipeline.
```

> The `ad-pipeline` sub-pipeline receives the metrics data from the entry pipeline and performs anomaly detection on the value of the metric using the [`anomaly_detector` processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/anomaly-detector/).

Reviewer 2: ```suggestion
The `ad-pipeline` sub-pipeline receives the metrics from the entry pipeline and performs anomaly detection on the value of the metric by using the [`anomaly_detector` processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/anomaly-detector/).
```

> The `ad-pipeline` sub-pipeline receives the metrics data from the entry pipeline and performs anomaly detection on the value of the metric using the [`anomaly_detector` processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/anomaly-detector/).

Reviewer 2: "and performs anomaly detection on the metric values"?
